mesonbuild/optinterpreter.py

```diff
@@ -166,60 +198,88 @@ def func_option(self, args: T.Tuple[str], kwargs: 'FuncOptionArgs') -> None:
         self.options[key] = opt
 
-    @permittedKwargs({'value', 'yield'})
-    def string_parser(self, description: str, kwargs: 'ParserArgs') -> coredata.UserOption:
-        value = kwargs.get('value', '')
-        return coredata.UserStringOption(description, value, kwargs['yield'])
-
-    @permittedKwargs({'value', 'yield'})
-    def boolean_parser(self, description: str, kwargs: 'ParserArgs') -> coredata.UserOption:
-        value = kwargs.get('value', True)
-        return coredata.UserBooleanOption(description, value, kwargs['yield'])
-
-    @permittedKwargs({'value', 'yield', 'choices'})
-    def combo_parser(self, description: str, kwargs: 'ParserArgs') -> coredata.UserOption:
-        choices = kwargs.get('choices')
-        if not choices:
-            raise OptionException('Combo option missing "choices" keyword.')
-        value = kwargs.get('value', choices[0])
-        return coredata.UserComboOption(description, choices, value, kwargs['yield'])
+    @typed_kwargs(
+        'string option',
+        KwargInfo('value', str, default=''),
+        _YIELD_KW,
+    )
+    def string_parser(self, description: str, kwargs: StringArgs) -> coredata.UserOption:
+        return coredata.UserStringOption(description, kwargs['value'], kwargs['yield'])
+
+    @typed_kwargs(
+        'boolean option',
+        KwargInfo(
+            'value',
+            (bool, str),
+            default=True,
+            validator=lambda x: None if isinstance(x, bool) or x in {'true', 'false'} else 'boolean options must have boolean values',
+        ),
+        _YIELD_KW,
+    )
+    def boolean_parser(self, description: str, kwargs: BooleanArgs) -> coredata.UserOption:
+        return coredata.UserBooleanOption(description, kwargs['value'], kwargs['yield'])
 
-    @permittedKwargs({'value', 'min', 'max', 'yield'})
-    def integer_parser(self, description: str, kwargs: 'ParserArgs') -> coredata.UserOption:
-        value = kwargs.get('value')
+    @typed_kwargs(
+        'combo option',
+        KwargInfo('value', (str, NoneType)),
+        KwargInfo('choices', ContainerTypeInfo(list, str, allow_empty=False), required=True),
+        _YIELD_KW,
+    )
+    def combo_parser(self, description: str, kwargs: ComboArgs) -> coredata.UserOption:
+        choices = kwargs['choices']
+        value = kwargs['value']
         if value is None:
-            raise OptionException('Integer option must contain value argument.')
-        inttuple = (kwargs.get('min'), kwargs.get('max'), value)
+            value = kwargs['choices'][0]
+        return coredata.UserComboOption(description, choices, value, kwargs['yield'])
+
+    @typed_kwargs(
+        'integer option',
+        KwargInfo(
+            'value',
+            (int, str),
+            default=True,
+            validator=lambda x: None if isinstance(x, int) or x.isdigit() else 'integer values must be either an integer or a string representation thereof',
```

Comment:
isdigit doesn't handle negative numbers. For non-string versions, we actually detect UMinusNode and extract its value, then feed it into func_option as `-arg` to handle this. For strings, we simply hand the string over to func_option and hit this code.